geode-core/src/main/java/org/apache/geode/internal/cache/wan/AbstractGatewaySenderEventProcessor.java

```diff
@@ -1322,17 +1325,180 @@ protected void afterExecute() {
 
   protected abstract void enqueueEvent(GatewayQueueEvent<?, ?> event);
 
-  private void pendingEventsInBatchesMarkAsPossibleDuplicate() {
+  private void notifyPossibleDuplicate(int reason, List<?> events) {
+    Map<String, Map<Integer, List<Object>>> regionToDispatchedKeysMap = new ConcurrentHashMap<>();
```

Comment:
Why `ConcurrentHashMap`? It doesn't seem to escape this stack/thread so where is the concurrency?

Reply:
Updated.